File: Java/백준/Bronze/2231. 분해합/분해합.java
import java.io.*;
import java.util.*;

public class Main {

    // 자릿수의 합을 계산하는 함수
    private static int digitSum(int number) {
        int sum = 0;
        while (number > 0) {
            sum += number % 10;
            number /= 10;
        }
        return sum;
    }

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        int result = 0;

        for (int i = Math.max(1, N - String.valueOf(N).length() * 9); i < N; i++) {
            int decompositionSum = i + digitSum(i);  // 분해합 계산
            if (decompositionSum == N) {
                result = i;
                break;
            }
        }

        System.out.println(result);
    }
}

File: Python/백준/Bronze/2231. 분해합/분해합.py
# 자릿수의 합을 계산하는 함수
def digit_sum(number):
    total = 0
    while number > 0: # 마지막 자리까지 반복
        total += number % 10
        number //= 10 # 1의 자리 버림
    return total


N = int(input())

result = 0
for i in range(max(1, N - len(str(N)) * 9), N):  # 범위 설정 (불필요한 연산을 방지하기 위해 분해합의 가능성이 있는 범위 산정)
    decomposition_sum = i + digit_sum(i)
    if decomposition_sum == N:
        result = i
        break

print(result)
